rfcs/0007-filtering.md

```diff
@@ -0,0 +1,93 @@
+# RFC Proposal
+[RFC Proposal]: #rfc-proposal
+
+* RFC Proposal ID: `0007-filtering-fields`
+* Start Date: 2020-08-21
+* RFC PR: [rfcs#0007](https://github.com/CMSgov/beneficiary-fhir-data/pull/345)
+* JIRA Ticket(s):
+    * [https://jira.cms.gov/browse/AB2D-1863](https://jira.cms.gov/browse/AB2D-1863)
+
+This proposal strengthens the guarantee made by the use of the _elements parameter for the **fhir/ExplanationOfBenefit** endpoint.
+
+## Table of Contents
+[Table of Contents]: #table-of-contents
+
+* [RFC Proposal](#rfc-proposal)
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Proposal](#proposal)
+    * [Proposed Solution: Unresolved Questions](#proposed-solution-unresolved-questions)
+    * [Proposed Solution: Drawbacks](#proposed-solution-drawbacks)
+    * [Proposed Solution: Notable Alternatives](#proposed-solution-notable-alternatives)
+* [Future Possibilities](#future-possibilities)
+
+## Motivation
+[Motivation]: #motivation
+
+A single BFD EOB retrieval call (UHC) with 2 million beneficiaries took 24 hours to retrieve three months of data.
+Larger PDPs can have in excess of 6 million beneficiaries.  To have a request take over 3 days is not acceptable.
+
+AB2D receives more data than needed from BFD when we make API calls to gather EOBs (Explanation of Benefits).  AB2D
+currently filters out irrelevant fields.  This might not be terrible for dozens of API calls, but for 
+millions of requests it can impact our performance significantly.  Since we only need a projection of the data
+(10 out of 69 top-level getters; 6 out of 42 fields in the ItemComponent), it makes
+sense for the server to send us only the data consumed.  This reduces bandwidth requirements for receiving the data.
```

Comment:
I would be interest in knowing what the actual bottleneck is.  When 2 million records are retrieved, it took 24 hours .. but how did that time spent break down?  If it was mostly due to transfer times then shrinking the output object would obviously help.  It could also help if serialization is the bottleneck.  If the bottleneck is the database retrieval or actual conversion from the native BFD data format to a FHIR EOB then other improvements would need to be applied.

From looking at the code a bit, it appears that the data retrieval and processing is all currently done serially.  If we could identify that as the main driver in this slower performance it may be easy to address without any API changes.

Reply:
We have built a testing framework specifically for this purpose and finally cleared the last configuration hurdle yesterday.  We will be experimenting in order to baseline the current peformance.  Once we have a baseline, AB2D is available for assistance in identifying additional bottlenecks.

In addition to shrinking the output object (saving on serialization and transmission times), any other performance improvements are welcome.  Work on improving database performance and conversion to FHIR EOB should be done independently of the effort to reduce payload size.

The proposal does not make any API changes, all functionality is covered by the current FHIR specification.  The proposal merely provides a higher service level guarantee than the FHIR specification.

Reply:
I agree that supporting `_elements` is a good idea in general, but it is being suggested as a fix for a performance issue.  We can't be sure that this would fix a performance issue without profiling BFD and identifying the root of the problem.  That was the only point I was making.